src/NUnitFramework/tests/Attributes/RangeAttributeTests.cs

```diff
@@ -22,478 +22,489 @@
+        public static void MaxValueRange_Single()
         {
-            CheckValuesWithinTolerance(nameof(MethodWithDoubleRangeAndNegativeStep_Reversed), 1.2, 1.0, 0.8);
+            Assert.That(
+                GetData(new RangeAttribute(float.MaxValue - SingleExtremaEpsilon * 2, float.MaxValue, SingleExtremaEpsilon), typeof(float)),
+                Is.EqualTo(new[] { float.MaxValue - SingleExtremaEpsilon * 2, float.MaxValue - SingleExtremaEpsilon, float.MaxValue }));
         }
 
-        private void MethodWithDoubleRangeAndNegativeStep_Reversed([Range(1.2, 0.7, -0.2)] double x) { }
-
         [Test]
-        public void DoubleRangeWithMultipleAttributes()
+        public static void MaxValueRangeLosingPrecision_Single()
         {
-            Test test = TestBuilder.MakeParameterizedMethodSuite(GetType(), nameof(MethodWithMultipleDoubleRange));
-
-            Assert.That(test.TestCaseCount, Is.EqualTo(6));
+            Assert.That(() => GetData(new RangeAttribute(float.MaxValue - SingleExtremaEpsilon, float.MaxValue, SingleExtremaEpsilon / 2), typeof(float)),
+                Throws.InstanceOf<ArithmeticException>());
         }
 
-        private void MethodWithMultipleDoubleRange([Range(1.0, 3.0, 1.0)] [Range(10.0, 12.0, 1.0)] double x) { }
-
-        #endregion
-
-        #region Floats
-
         [Test]
-        public void FloatRangeAndStep()
+        public static void MaxValueRange_Double()
         {
-            CheckValuesWithinTolerance(nameof(MethodWithFloatRangeAndStep), 0.7f, 0.9f, 1.1f);
+            Assert.That(
+                GetData(new RangeAttribute(double.MaxValue - DoubleExtremaEpsilon * 2, double.MaxValue, DoubleExtremaEpsilon), typeof(double)),
+                Is.EqualTo(new[] { double.MaxValue - DoubleExtremaEpsilon * 2, double.MaxValue - DoubleExtremaEpsilon, double.MaxValue }));
         }
 
-        private void MethodWithFloatRangeAndStep([Range(0.7f, 1.2f, 0.2f)] float x) { }
-
         [Test]
-        public void FloatRangeAndZeroStep()
+        public static void MaxValueRangeLosingPrecision_Double()
         {
-            Assert.Throws<ArgumentException>(() => CheckValuesWithinTolerance(nameof(MethodWithFloatRangeAndZeroStep), 0.7f, 0.9f, 1.1f));
+            Assert.That(() => GetData(new RangeAttribute(double.MaxValue - DoubleExtremaEpsilon, double.MaxValue, DoubleExtremaEpsilon / 2), typeof(double)),
+                Throws.InstanceOf<ArithmeticException>());
         }
 
-        private void MethodWithFloatRangeAndZeroStep([Range(0.7f, 1.2f, 0.0f)] float x) { }
-
         [Test]
-        public void FloatRangeAndStep_Reversed()
+        public static void MaxValueRange_Decimal()
         {
-            Assert.Throws<ArgumentException>(() => CheckValuesWithinTolerance(nameof(MethodWithFloatRangeAndStep_Reversed), 0.7f, 0.9f, 1.1f));
-        }
+            const double step = (double)(((decimal)LargestDoubleConvertibleToDecimal - (decimal)NextLargestDoubleConvertibleToLowerDecimal) / 2);
```

Comment:
If we extracted `(decimal)NextLargestDoubleConvertibleToLowerDecimal` into a variable (with a shorter name) then perhaps the test would be easier to read?

Reply:
I ended up doing something similar. Does https://github.com/nunit/nunit/pull/2805/commits/e5cc4c8d26b42ee8d50efeaf7e5f5615deb4ab89 make it easy to read?